### DISCUSSION TITLE
Services: RADICAL+SOMA Performance Monitoring

### DISCUSSION THREAD

**Discussion by mtitov** [radical-cybertools/radical.pilot#3006] [General]
SOMA repository: https://github.com/soma-monitoring-toolbox

**Actionable items**
- Extend support of service tasks (startup flag + termination) 
  - Ensure having a comprehensive set of tests for service tasks
- Service/daemon to collect information about resource usage (per node)
- Service/daemon to collect information about task state (e.g., tracing events)

Collected information is pushed to the SOMA service using RPC calls.

**Related tickets**

_Prioritized tickets_
   - https://github.com/radical-cybertools/radical.pilot/issues/3007
   - https://github.com/radical-cybertools/radical.pilot/issues/3008

_Other tickets_
- Service startup flag: https://github.com/radical-cybertools/radical.pilot/issues/2953
  - As a temporary approach - consider using "readiness" file - when service is up it creates a corresponding file, and RP waits for this file before proceeding further 
- Service termination: https://github.com/radical-cybertools/radical.pilot/issues/2967
- https://github.com/radical-cybertools/radical.pilot/issues/2899
- https://github.com/radical-cybertools/radical.pilot/issues/2543
- https://github.com/radical-cybertools/radical.pilot/issues/2711

p.s. for the reference, previous proposal for monitoring: https://github.com/radical-cybertools/radical.pilot/discussions/2308

**Comment by mtitov** [comment]
First step: implement component for task monitoring ([original comment](https://github.com/radical-cybertools/radical.pilot/issues/3007#issuecomment-1698112820))

Next steps to build service/daemon to collect information about resources usage
- [ ] remove sub-agents
- [ ] place components directly on target nodes
- [ ] implement service as special components

**Comment by andre-merzky** [comment]
After discussion we plan to implement the following schema

  - services are defined as normal RP tasks (special task type though)
  - service tasks are tagged to a specific pilot instance by setting the `pilot` field in the TD
  - they are submitted via the `tmgr`
  - we introduce a new control message which signals the event that a service task is `ready` (which implies it has reached `AGENT_EXECUTING` state.  That message type will have the `fwd` flag set by default
  - the `tmgr` will collect those events and will issue `ready`-callbacks which are definable on service task types
